File: API/src/main/java/com/ltizzi/ecommerce/model/purchase/PurchaseEntity.java
package com.ltizzi.ecommerce.model.purchase;

import com.ltizzi.ecommerce.model.shoporder.ShopOrderEntity;
import com.ltizzi.ecommerce.model.user.UserEntity;
import jakarta.persistence.*;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.annotations.Where;

import java.math.BigDecimal;
import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.List;

/**
 * @author Leonardo Terlizzi
 */
@Data
@NoArgsConstructor
@Entity
@Table(name = "purchases")
@SQLDelete(sql = "UPDATE purchases SET soft_delete purchase_id=?")
@Where(clause = "soft_delete = false")
public class PurchaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long purchase_id;

    @OneToMany(fetch = FetchType.LAZY)
//    @JoinColumn(name = "shop_order_id")
    private List<ShopOrderEntity> orders = new ArrayList<>();

    private BigDecimal total_income;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private UserEntity user;

    @CreationTimestamp
    @Column(updatable = false)
    private Timestamp createdAt;

    @UpdateTimestamp
    private Timestamp updatedAt;

    @Column(name = "soft_delete")
    private Boolean soft_delete = Boolean.FALSE;
}

File: Server/src/entities/PurchaseEntity.ts
import {
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  JoinTable,
  ManyToMany,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from "typeorm";
import { Purchase } from "../models/Purchase";
import { ShopOrder } from "../models/ShopOrder";
import { User } from "../models/User";
import { ShopOrderEntity } from "./ShopOrderEntity";
import { UserEntity } from "./UserEntity";

@Entity("purchases")
export class PurchaseEntity implements Purchase {
  @PrimaryGeneratedColumn()
  purchase_id!: number;

  @ManyToMany(
    () => ShopOrderEntity,
    { eager: true, nullable: false }

    // {cascade: true,
    // eager: true,
    // nullable: false,
    // onDelete: "CASCADE",}
  )
  // @JoinColumn({ name: "orders", referencedColumnName: "shop_order_id" })
  @JoinTable()
  orders!: ShopOrder[];

  @Column({ type: "numeric" })
  total_income!: number;

  @Column()
  user_id!: number;

  @ManyToOne(() => UserEntity, (user) => user.purchases, {
    eager: false,
    nullable: false,
  })
  @JoinColumn({ name: "user_id", referencedColumnName: "user_id" })
  user!: User;

  @CreateDateColumn({
    type: "timestamp",
    default: () => "CURRENT_TIMESTAMP",
  })
  createdAt!: Date;

  @UpdateDateColumn({
    type: "timestamp",
    default: () => "CURRENT_TIMESTAMP",
    onUpdate: "CURRENT_TIMESTAMP",
  })
  updatedAt!: Date;

  @Column({ default: false })
  soft_delete!: boolean;
}
